compiler/rustc_mir/src/transform/large_enums.rs

```diff
@@ -0,0 +1,278 @@
+use crate::transform::MirPass;
+use rustc_data_structures::stable_map::FxHashMap;
+use rustc_middle::mir::*;
+use rustc_middle::ty::util::IntTypeExt;
+use rustc_middle::ty::{self, Const, Ty, TyCtxt};
+use rustc_span::def_id::DefId;
+use rustc_target::abi::{HasDataLayout, Size, TagEncoding, Variants};
+use std::array::IntoIter;
+
+/// A pass that seeks to optimize unnecessary moves of large enum types, if there is a large
+/// enough discrepanc between them
+pub struct EnumSizeOpt<const DISCREPANCY: u64>;
+
+impl<'tcx, const D: u64> MirPass<'tcx> for EnumSizeOpt<D> {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        self.optim(tcx, body);
+    }
+}
+
+impl<const D: u64> EnumSizeOpt<D> {
+    fn candidate<'tcx>(
+        tcx: TyCtxt<'tcx>,
+        ty: Ty<'tcx>,
+        body_did: DefId,
+    ) -> Option<(u64, Vec<Size>)> {
+        match ty.kind() {
+            ty::Adt(adt_def, _substs) if adt_def.is_enum() => {
+                let p_e = tcx.param_env(body_did);
```

Comment:
The `param_env` would be a conventional name, instead of `p_e`. I would also call `tcx.param_env(body_did)` only once at the start of the pass.